src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,987 @@ mod v1 {
+            let mut total_size = 0;
+            for cf_file in &self.cf_files {
+                delete_file_if_exist(&cf_file.tmp_path);
+                delete_file_if_exist(&cf_file.path);
+                if exists {
+                    total_size += cf_file.size;
+                }
+            }
+            delete_file_if_exist(&self.meta_file.tmp_path);
+            delete_file_if_exist(&self.meta_file.path);
+            if exists {
+                let mut size_track = self.size_track.wl();
+                *size_track = size_track.saturating_sub(total_size);
+            }
+        }
+
+        fn validate(&self) -> RaftStoreResult<()> {
+            for cf_file in &self.cf_files {
+                if cf_file.size == 0 {
+                    // Skip empty file. The checksum of this cf file should be 0 and
+                    // this is checked when loading the snapshot meta.
+                    continue;
+                }
+                try!(check_file_size_and_checksum(&cf_file.path, cf_file.size, cf_file.checksum));
+            }
+            Ok(())
+        }
+
+        fn switch_to_cf_file(&mut self, cf: &str) -> io::Result<()> {
+            match self.cf_files.iter().position(|x| x.cf == cf) {
+                Some(index) => {
+                    self.cf_index = index;
+                    Ok(())
+                }
+                None => Err(io::Error::new(ErrorKind::Other, format!("fail to find cf {}", cf))),
+            }
+        }
+
+        fn add_kv(&mut self, k: &[u8], v: &[u8]) -> io::Result<()> {
+            let mut cf_file = &mut self.cf_files[self.cf_index];
+            let mut writer = cf_file.sst_writer.as_mut().unwrap();
+            if let Err(e) = writer.add(k, v) {
+                return Err(io::Error::new(ErrorKind::Other, e));
+            }
+            cf_file.kv_count += 1;
+            Ok(())
+        }
+
+        fn save_cf_files(&mut self) -> io::Result<()> {
+            for cf_file in &mut self.cf_files {
+                if plain_file_used(cf_file.cf) {
+                    let _ = cf_file.file.take();
+                } else if cf_file.kv_count == 0 {
+                    let _ = cf_file.sst_writer.take().unwrap();
+                } else {
+                    let mut writer = cf_file.sst_writer.take().unwrap();
+                    if let Err(e) = writer.finish() {
+                        return Err(io::Error::new(ErrorKind::Other, e));
+                    }
+                }
+                let size = try!(get_file_size(&cf_file.tmp_path));
+                if size > 0 {
+                    try!(fs::rename(&cf_file.tmp_path, &cf_file.path));
+                    cf_file.size = size;
+                    cf_file.checksum = try!(calc_crc32(&cf_file.path));
+                } else {
+                    // Clean up the `tmp_path` if this cf file is empty.
+                    delete_file_if_exist(&cf_file.tmp_path);
```

Comment:
I thought `Drop` was not implemented when I saw the temp file was deleted explicitly here.